bundles/org.openhab.binding.sensorpush/src/main/java/org/openhab/binding/sensorpush/internal/SensorPushDiscoveryService.java

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.sensorpush.internal;
+
+import static org.openhab.binding.sensorpush.internal.SensorPushBindingConstants.*;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.sensorpush.internal.handler.CloudBridgeHandler;
+import org.openhab.binding.sensorpush.internal.protocol.Sensor;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResult;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.ThingHandler;
+import org.openhab.core.thing.binding.ThingHandlerService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link SensorPushDiscoveryService} handles discovery of sensors as they are identified by the bridge handler.
+ * Requests from the framework to startScan() just clear the discovered sensor set. The next scheduled poll will
+ * re-discover all sensors.
+ *
+ * @author Bob Adair - Initial contribution
+ */
+@NonNullByDefault
+public class SensorPushDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(SensorPushDiscoveryService.class);
+
+    private @NonNullByDefault({}) CloudBridgeHandler bridgeHandler;
+    private final Set<String> discoveredSensorSet = new HashSet<>();
+
+    public SensorPushDiscoveryService() {
+        super(DISCOVERABLE_DEVICE_TYPE_UIDS, 0, false);
+    }
+
+    @Override
+    protected void startScan() {
+        logger.trace("Starting discovery scan");
+        discoveredSensorSet.clear();
```

Comment:
The framework takes care that already discovered Things don't get added again. You only need to define a representation property, which you already did.

You could trigger the polling here.

Reply:
Yes, originally I had intended to trigger a poll from here, but I was worried about potentially running in to problems with API call rate limiting on the cloud service because of the extra unscheduled poll, so I decided to leave it out for now. I'll probably add it in the future.

Reply:
I think it's ok if the next normal poll fails due to rate limiting after the user hit "start scanning". He/she expects to start a scan if "start scan" is clicked. I leave it up to you how to handle that.

In any case, the discovered sensor map could be removed.

Reply:
Ok. I updated the startScan() method to call the bridge handler's pollSensors() method. I left discoveredSensorSet in there, though, because otherwise the discovery code will be hit for every sensor every other poll interval, which is pretty frequently.